common/src/main/java/io/pravega/common/io/FileModificationWatcher.java

```diff
@@ -0,0 +1,126 @@
+ * - If there are multiple file modifications in quick succession, only the last one may trigger the specified action.
+ *
+ */
+@Slf4j
+public class FileModificationWatcher extends Thread {
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the specified file changes.
+     */
+    private final Consumer<WatchEvent<?>> callback;
+
+    private final UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> handleException(t.getName(), e);
+
+    public FileModificationWatcher(@NonNull String fileToWatch, @NonNull Consumer<WatchEvent<?>> callback) {
+        super();
+
+        this.pathOfFileToWatch = Paths.get(fileToWatch);
+
+        if (!this.pathOfFileToWatch.toFile().exists()) {
+            throw new IllegalArgumentException(String.format("File [%s] does not exist.", fileToWatch));
+        }
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service.");
+
+            String fileName = this.pathOfFileToWatch.getFileName().toString();
+            Path directoryPath = this.pathOfFileToWatch.getParent();
+
+            log.debug("Directory being watched is {}.", directoryPath);
+
+            directoryPath.register(watchService,
+                    StandardWatchEventKinds.ENTRY_MODIFY);
+            log.debug("Done setting up watch for modify entries.");
+
+            while (true) {
+                watchKey = watchService.take();
+                log.debug("Done setting up watch key.");
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
+
+                if (watchKey != null) {
+                    log.debug("Watch key is not null");
+
+                    watchKey.pollEvents()
+                            .stream()
+                            .filter(event -> // we only care about changes to the specified file.
+                                    event.context().toString().contains(fileName))
+                            .forEach(event ->  callback.accept(event)); // invoke the specified callback
+                }
+
+                boolean isKeyValid = watchKey.reset();
+                log.debug("Done resetting watch key, so that it can receive further event notifications.");
+                if (!isKeyValid) {
+                    log.info("No longer watching file [{}]", this.pathOfFileToWatch);
```

Comment:
So how do you exit this while loop? I would have expected this to break from it, otherwise it would keep on looping forever.

Reply:
Sure, I've modified the code to exit the loop at this point.